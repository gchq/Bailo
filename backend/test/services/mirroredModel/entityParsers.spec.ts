import { describe, expect, test, vi } from 'vitest'

import { parseFile, parseModelCard, parseRelease } from '../../../src/services/mirroredModel/entityParsers.js'

const s3Mocks = vi.hoisted(() => ({
  objectExists: vi.fn(() => Promise.resolve(true)),
}))
vi.mock('../../../src/clients/s3.js', () => s3Mocks)

const mockLogData = { extra: 'info', importId: 'importId' }

describe('services > parsers > modelParser', () => {
  test('parseModelCard > success', () => {
    const result = parseModelCard(
      {
        modelId: 'sourceModelId',
        schemaId: 'schemaId',
        version: 2,
        createdBy: 'user',
        updatedAt: 'timestamp',
        createdAt: 'timestamp',
        _id: 'mongoId',
      },
      'mirroredModelId',
      'sourceModelId',
      mockLogData,
    )

    expect(result).toMatchSnapshot()
  })

  test('parseModelCard > data not a model card', () => {
    expect(() => parseModelCard({}, '', '', mockLogData)).toThrowError('Data cannot be converted into a model card.')
  })

  test('parseModelCard > bad sourceModelId', () => {
    expect(() =>
      parseModelCard(
        {
          modelId: 'sourceModelId',
          schemaId: 'schemaId',
          version: 2,
          createdBy: 'user',
          updatedAt: 'timestamp',
          createdAt: 'timestamp',
          _id: 'mongoId',
        },
        'mirroredModelId',
        'badSourceModelId',
        mockLogData,
      ),
    ).toThrowError('Compressed file contains model cards that have a model ID that does not match the source model Id.')
  })

  test('parseRelease > success', () => {
    const result = parseRelease(
      {
        modelId: 'sourceModelId',
        modelCardVersion: 1,
        semver: '0.0.1',
        notes: 'notes',
        minor: false,
        draft: false,
        fileIds: [],
        images: [],
        deleted: false,
        createdBy: 'user',
        updatedAt: 'timestamp',
        createdAt: 'timestamp',
        _id: 'mongoId',
      },
      'mirroredModelId',
      'sourceModelId',
      mockLogData,
    )

    expect(result).toMatchSnapshot()
  })

  test('parseRelease > data not a release', () => {
    expect(() => parseRelease({}, '', '', mockLogData)).toThrowError('Data cannot be converted into a release.')
  })

  test('parseRelease > bad sourceModelId', () => {
    expect(() =>
      parseRelease(
        {
          modelId: 'sourceModelId',
          modelCardVersion: 1,
          semver: '0.0.1',
          notes: 'notes',
          minor: false,
          draft: false,
          fileIds: [],
          images: [],
          deleted: false,
          createdBy: 'user',
          updatedAt: 'timestamp',
          createdAt: 'timestamp',
          _id: 'mongoId',
        },
        'mirroredModelId',
        'badSourceModelId',
        mockLogData,
      ),
    ).toThrowError('Compressed file contains releases that have a model ID that does not match the source model Id.')
  })

  test('parseFile > success', () => {
    const result = parseFile(
      {
        modelId: 'sourceModelId',
        name: 'file.txt',
        size: 123,
        mime: 'text/plain',
        path: 'foo/file.txt',
        complete: true,
        deleted: false,
        updatedAt: 'timestamp',
        createdAt: 'timestamp',
        _id: 'mongoId',
      },
      'mirroredModelId',
      'sourceModelId',
      mockLogData,
    )

    expect(result).toMatchSnapshot()
  })

  test('parseFile > data not a file', async () => {
    await expect(() => parseFile({}, '', '', mockLogData)).rejects.toThrowError('Data cannot be converted into a file.')
  })

  test('parseFile > file does not exist', async () => {
    s3Mocks.objectExists.mockRejectedValueOnce('Error')
    await expect(() =>
      parseFile(
        {
          modelId: 'sourceModelId',
          name: 'file.txt',
          size: 123,
          mime: 'text/plain',
          path: 'foo/file.txt',
          complete: true,
          deleted: false,
          updatedAt: 'timestamp',
          createdAt: 'timestamp',
          _id: 'mongoId',
        },
        'mirroredModelId',
        'sourceModelId',
        mockLogData,
      ),
    ).rejects.toThrowError('Error checking existence of file in storage.')
  })

  test('parseFile > bad sourceModelId', async () => {
    await expect(() =>
      parseFile(
        {
          modelId: 'sourceModelId',
          name: 'file.txt',
          size: 123,
          mime: 'text/plain',
          path: 'foo/file.txt',
          complete: true,
          deleted: false,
          updatedAt: 'timestamp',
          createdAt: 'timestamp',
          _id: 'mongoId',
        },
        'mirroredModelId',
        'badSourceModelId',
        mockLogData,
      ),
    ).rejects.toThrowError(
      'Compressed file contains files that have a model ID that does not match the source model Id.',
    )
  })
})
