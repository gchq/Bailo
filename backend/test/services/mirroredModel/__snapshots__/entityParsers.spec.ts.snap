// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`services > parsers > modelParser > parseFile > success 1`] = `Promise {}`;

exports[`services > parsers > modelParser > parseModelCard > success 1`] = `
{
  "createdAt": "timestamp",
  "createdBy": "user",
  "modelId": "mirroredModelId",
  "schemaId": "schemaId",
  "updatedAt": "timestamp",
  "version": 2,
}
`;

exports[`services > parsers > modelParser > parseRelease > success 1`] = `
{
  "createdAt": "timestamp",
  "createdBy": "user",
  "deleted": false,
  "draft": false,
  "fileIds": [],
  "images": [],
  "minor": false,
  "modelCardVersion": 1,
  "modelId": "mirroredModelId",
  "notes": "notes",
  "semver": "0.0.1",
  "updatedAt": "timestamp",
}
`;
